rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,315 @@
+}
+```
+
+<details>
+<summary>Input / Output</summary>
+
+```javascript 
+// input code
+
+for (const el of arr) {
+  el.logger?.log("Ok!");
+}
+```
+
+```javascript 
+// output code
+
+for (let i = 0; i < arr.length; i++) {
+  const el = arr[i];
+  el.logger == null ? void 0 : el.logger.log("Ok!");
+}
+```
+</details>
+<br />
+
+```jsonc
+// babel.config.json
+{
+  "assumptions": {
+    "noDocumentAll": false,
+    "noGetterSideEffects": true,
+    "iterableIsArray": true
+  },
+  "presets": ["@babel/preset-env"]
+}
+```
+
+
+<details>
+<summary>Input / Output</summary>
+
+```javascript 
+// input code
+
+for (const el of arr) {
+  el.logger?.log("Ok!");
+}
+```
+
+```javascript 
+// output code
+
+for (let i = 0; i < arr.length; i++) {
+  const el = arr[i];
+  el.logger === null || el.logger === void 0 ? void 0 : el.logger.log("Ok!");
+}
+```
+</details>
+
+# Motivation
+
+The `loose` options have different problems:
+- _"What is loose?"_ Since `loose` is not descriptive, it requires users to check a plugin's docs to see what assumptions it makes (and the docs often only partially mention what `loose` means in a specific context).
+- _"How loose is loose?"_ Is making `loose` "looser" a breaking change? We have been considering it as such: for this reason, we have been introducing different `loose`-like options in some plugins. For example, `transform-for-of` supports `loose`, `assumeArray` and `allowArrayLike`.
+- There are cross-dependencies between the `loose` option of different plugins: for example, `proposal-class-properties`'s loose mode must always match `proposal-private-methods`'s (otherwise we throw an error). This has already caused problems: [babel/babel#11622](https://github.com/babel/babel/issues/11622).
+- Some plugins should be aware of other plugins' `loose` option value. For example, we sometimes need to partially transpile optional chaining in the class properties plugin ([babel/babel#11248](https://github.com/babel/babel/pull/11248)). Should we add a new `optionalChainingLoose` option to `proposal-private-methods` and `proposal-class-properties`? Or should we use a cross-plugin communication channel to share it? What about when someone has the private methods plugin enabled, but not the optional chaining one?
+
+This RFC also solves another problem, which is not directly related to the points above:
+- `@babel/preset-env` has a `loose` option, which is forwarded to all the plugins it enables. When a user wants to set `loose` only for a specific feature, they have to manually install the plugin and explicitly add it to their configuration ([babel/babel#6978](https://github.com/babel/babel/issues/6978)).
+
+
+# Detailed design
+
+## New configuration option
+
+This RFC introduces a new top-level option: `assumptions`. This is an object containing different flags to mark different assumptions about the input code as safe and thus allow optimizing them.
+
+```typescript
+type Assumptions = {
+  [assumption: string]: boolean
```

Comment:
Yeah I described them as simple booleans exactly to avoid thinking about merging them :sweat_smile: 

I feel like we can just use longer names instead of nesting objects.